### PR TITLE
feat: MRMFeatureSelector

### PR DESCRIPTION
**Pull request by pcolaianni**
I would like to get some feedback about the state of the classes `MRMFeatureScheduler` and `MRMFeatureSelector`.

The documentation part could use some inspiration :smile: It's the part that probably needs most help.

How would you call the method which constructs what was the `ToList` in the python version? It is currently named `constructToList()`

### CODE REVIEW

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    bool   locality_weight_         = false;
+    bool   select_transition_group_ = true;
+    Int    segment_window_length_   = 8;
+    Int    segment_step_length_     = 4;
+    String variable_type_           = "continuous";
+    double optimal_threshold_       = 0.5;
+    std::map<String, String> score_weights_;
+
+    // TODO: rename method
+    void constructToList(
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h]:
```suggestion
    void constructTargTransList(
```

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    Int    nn_threshold_            = 4;
+    bool   locality_weight_         = false;
+    bool   select_transition_group_ = true;
+    Int    segment_window_length_   = 8;
+    Int    segment_step_length_     = 4;
+    String variable_type_           = "continuous";
+    double optimal_threshold_       = 0.5;
+    std::map<String, String> score_weights_;
+
+    // TODO: rename method
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h]:
```suggestion
    /*
    @brief Construct the target transition or transition group retention times that will be used
    to score candidate features based on their deviation from the 
    relative distance between the target transition or transition group times
    */
```

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  {
+public:
+    void optimize(
+      const std::vector<std::pair<double, String>>& time_to_name,
+      const std::map< String, std::vector<Feature> >& feature_name_map,
+      std::vector<String>& result
+    );
+  };
+
+  class MRMFeatureSelector_test : public MRMFeatureSelectorQMIP
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h:298]: This class should be moved to the tests correct?

File: src/tests/class_tests/openms/source/MRMFeatureSelector_test.cpp [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+  score = selector.compute_score(feature);
+  TEST_REAL_SIMILAR(score, 0.20059549093267626)
+
+  selector.setScoreWeights({{"sn_ratio", "lambda score: 1/log(score)"}, {"peak_apices_sum", "lambda score: 1/log10(score)"}});
+  score = selector.compute_score(feature);
+  TEST_REAL_SIMILAR(score, 0.10427624775717449)
+}
+END_SECTION
+
+// START_SECTION(schedule_MRMFeaturesQMIP() integer) // integer variable type
```
Comment (dmccloskey) [inline src/tests/class_tests/openms/source/MRMFeatureSelector_test.cpp]: I would recommend to leave this in there.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureScheduler.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+#include <OpenMS/DATASTRUCTURES/Param.h>
+#include <OpenMS/DATASTRUCTURES/DefaultParamHandler.h>
+#include <OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h>
+#include <OpenMS/KERNEL/FeatureMap.h>
+#include <map>
+#include <vector>
+
+namespace OpenMS
+{
+  class OPENMS_DLLAPI MRMFeatureScheduler
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureScheduler.h:52]: Docstrings: "Class used to 'schedule' multiple calls to `MRMFeatureSelector`"

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    ) const;
+
+    /// Transform the given score thorugh the chosen lambda function
+    double weight_func(const double score, const String& lambda_score) const;
+
+    /// Helper method to remove spaces from a string
+    String remove_spaces(String str) const;
+  };
+
+  class OPENMS_DLLAPI MRMFeatureSelectorQMIP : public MRMFeatureSelector
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h:253]: Class used to select MRMFeatures based on relative retention time using a quadratic mixed integer programming (QMIP) formulation.  The method is described in [TODO: update when published]

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+  {
+public:
+    void optimize(
+      const std::vector<std::pair<double, String>>& time_to_name,
+      const std::map< String, std::vector<Feature> >& feature_name_map,
+      std::vector<String>& result
+    );
+  };
+
+  class OPENMS_DLLAPI MRMFeatureSelectorScore : public MRMFeatureSelector
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h:281]: Class used to select MRMFeatures based on a linear programming where each possible transition is weighted by a user defined score (most often retention time and peak intensity). The method is described in [TODO: update when published].

**Review by dmccloskey — CHANGES_REQUESTED** [review]
This looks really nice!  Great job with the conversion.  The cpp implementation looks much much better than my py implementation.

Once the comments/suggestions that I made are implemented, it should be ready for a PR.

File: src/tests/class_tests/openms/source/MRMFeatureSelector_test.cpp (hunk 4 above)
Comment (pcolaianni) [reply src/tests/class_tests/openms/source/MRMFeatureSelector_test.cpp]: Meaning, leave it as a commented-out piece of code?
The reason I commented it out is the time it takes to process (being in the order of minutes).

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h (hunk 3 above)
Comment (pcolaianni) [reply src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h:298]: Not sure that I can do that.
Anyway, we have been keeping those test `friend` classes within the same .h file for the other past classes.

File: src/tests/class_tests/openms/source/MRMFeatureSelector_test.cpp (hunk 4 above)
Comment (dmccloskey) [reply src/tests/class_tests/openms/source/MRMFeatureSelector_test.cpp]: I see.  OK just leave it as a comment then.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h (hunk 3 above)
Comment (dmccloskey) [reply src/openms/include/OpenMS/ANALYSIS/OPENSWATH/MRMFeatureSelector.h:298]: Leave as is then.

**Pull request closed without merging**